commons/Common.ml

```diff
@@ -1140,6 +1149,28 @@ let erase_this_temp_file f =
 (* List *)
 (*****************************************************************************)
 
+(* Safe implementation that prevents stack overflows. *)
+let map f xs =
+  (* Since map is such a frequently used function we try to make it fast for
+   * small/medium-sized lists. Inspired by Jane Street's Base library. *)
+  let max_rec_count = 1000 in
+  let rec count_map i = function
+    | [] -> []
+    | x1::[] -> f x1 :: []
+    | x1::x2::[] -> f x1 :: f x2 :: []
+    | x1::x2::x3::xs ->
+        let ys =
+          if i <= max_rec_count then
+            count_map (i+1) xs
+          else
+            xs
+            |> List.rev_map f
+            |> List.rev
+        in
+        f x1 :: f x2 :: f x3 :: ys
+  in
+  count_map 0 xs
```

Comment:
This implementation doesn't guarantee that `f` is applied from left to right. I think we should offer this guarantee (just use `let ... in`) and document it. That would be another nice guarantee that the default `List.map` doesn't offer.

Reply:
Ooops, that's a good catch! We should definitely fix that. Side-effects...